### PR TITLE
[ui] Allow searching jobs by Run ID

### LINKED ISSUE #15260: Search by run id in job history
### What's the use case?

In many cases we get run id from logging system and want to locate the run, but in the new Dagster UI it doesn't seem to have a(intuitive) way to search:
example

### Ideas of implementation

Maybe add an extra menu item "Run Id"?

### Additional information

_No response_

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

### PR DESCRIPTION
**Pull request by bengotow**
## Summary & Motivation

This is a small PR that makes it possible to filter runs by Run ID (Fixes #15260).

We mostly already supported this, but there wasn't dropdown UI for the filter.  This filter only allows you to enter 36-character Run IDs in the UUID format, which is unfortunate because the UI shows the 8 character version. It seems like it'd actually take a significant amount of work to allow "prefix-" matching against the 8-character display IDs because this filter goes all the way down to the RunStorage interface and there are multiple implementations.

To make this limitation more clear, I hacked in an "empty state" which explains what you need to provide. (Screenshot 2), and changed the `freeformSearchResult` callback so that it can return null to offer no completion if the user's input is not a valid value.

While I was in here, I also created local vars in the RunsFilterInput for some of the fillers that were inlined so the code is a bit more consistent. 

image

image

image

## How I Tested These Changes

I tested this by filtering for run IDs, pasting / typing various text and reloading the pages to make sure the filter persists.

### CODE REVIEW

File: js_modules/dagster-ui/packages/ui-core/src/runs/RunsFilterInput.tsx [hunk 1]
```diff
@@ -395,166 +393,214 @@ export const useRunsFilterInput = ({tokens, onChange, enabledFilters}: RunsFilte
...
-          if (value.type === DagsterTag.SensorName) {
-            icon = <Icon name="sensors" />;
-          } else if (value.type === DagsterTag.ScheduleName) {
-            icon = <Icon name="schedule" />;
-          } else if (value.type === DagsterTag.User) {
-            return <UserDisplay email={value.value!} isFilter />;
-          } else if (value.type === DagsterTag.Automaterialize) {
-            icon = <Icon name="auto_materialize_policy" />;
-            labelValue = 'Auto-materialize policy';
+  const launchedByFilter = useStaticSetFilter({
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/runs/RunsFilterInput.tsx:396]: This is unchanged, I just moved it to a local variable like the others

File: js_modules/dagster-ui/packages/ui-core/src/runs/RunsFilterInput.tsx [hunk 2]
```diff
@@ -395,166 +393,214 @@ export const useRunsFilterInput = ({tokens, onChange, enabledFilters}: RunsFilte
...
+          values[1] != null ? {token: 'created_date_before', value: `${values[1] / 1000}`} : null,
+        ].filter((x) => x) as RunFilterToken[]),
+      ]);
+    },
+  });
+
+  const ID_EMPTY = 'Type or paste 36-character ID';
+  const ID_TOO_SHORT = 'Invalid Run ID';
+
+  const idFilter = useSuggestionFilter({
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/runs/RunsFilterInput.tsx:479]: This `idFilter` is the only new code here!

File: js_modules/dagster-ui/packages/ui-core/src/runs/RunsFilterInput.tsx [hunk 3]
```diff
@@ -395,166 +393,214 @@ export const useRunsFilterInput = ({tokens, onChange, enabledFilters}: RunsFilte
...
+    name: 'Run ID',
+    icon: 'id',
+    initialSuggestions: [{final: true, value: ID_EMPTY}],
+    state: React.useMemo(() => {
+      return tokens.filter(({token}) => token === 'id').map((token) => token.value);
+    }, [tokens]),
+    freeformSearchResult: (query) => {
+      return /^([a-f0-9-]{36})$/.test(query.trim()) ? {value: query.trim(), final: true} : null;
+    },
```
Comment (salazarm) [inline js_modules/dagster-ui/packages/ui-core/src/runs/RunsFilterInput.tsx:488]: We should add an "emptyPlaceholder" prop to the base filter

**Review by salazarm — APPROVED** [review]

File: js_modules/dagster-ui/packages/ui-core/src/runs/RunsFilterInput.tsx (hunk 3 above)
Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/runs/RunsFilterInput.tsx:488]: to make it work I would make FilterDropdown.tsx call getPlaceholder() or something like that on the `selectedFilter` and pass it the current search query. The filter can return null if they don't want to show a placeholder or return a placeholder. That way we don't make the message clickable.

Comment (bengotow) [reply js_modules/dagster-ui/packages/ui-core/src/runs/RunsFilterInput.tsx:488]: This sounds good @salazarm - will make this change!

**Pull request merged**